Syntax highlighting support

Juliotati opened the pull request:
- Fixes #51 
- Fixes #43 

<img width="669" alt="Screenshot 2023-05-19 at 14 53 36" src="https://github.com/mkobuolys/flutter-design-patterns/assets/59662912/ad66c2c4-eaab-45ef-837f-9d0f2cb5d287">

Juliotati commented:
@mkobuolys Before i continue can you take a look and let me know if you're okay with the current result? 👍
I've only applied changes to one markdown file `abstract-factory.md` as you can see on the screenshot.

A richer highlighting will requires further contributions on [highlight](https://pub.dev/packages/highlight) package for the dart lang file `dart.dart`, it's missing the `late` and the new feature we got on dart.

mkobuolys commented:
Hey, thanks for your contribution! I have several things in mind that seems off and I am not a fan of:

1. All the code examples have the "dart" keyword in the first line. Probably, `parseSyntaxLanguage` is responsible for that, isn't it?
2. Headers in the Markdown are way too big.
3. The purple colour to highlight code lines in the Markdown seems random.
4.  I am not very happy with the code quality overall. Not that it's bad, just it feels inconsistent with the remaining code in the project:
  4.1. There are linter warnings.
  4.2. Using a Stateful widget for the `RichMarkdownBody` together with the `AutomaticKeepAliveClientMixin` seems extra - do you have any arguments for why it is needed?
  4.3.  try/catch block for the Markdown file is not needed - all the Markdown files in the project are pre-defined. Thus, we should not cover the "cannot render" cases but rather fix the Markdown file to make it work.
  4.4. In the project, there is a `LayoutConstants` file for spacing constants - these should be used from there. Also, I try to use values of 2 multipliers (2, 4, 8, 16, 32) overall.
  4.5. I am not a fan of global functions - it seems that they could just live inside the corresponding Widget classes.
  4.6. I would extract the `_syntaxTheme` to a dedicated file with defined constant colour values and so on.
  
That's what I could think of for now. If there are any concerns or questions - let me know.

Juliotati commented:
> 1. Fixed ✅ (At the cost of only recognising **dart** syntax, but i guess it's fine for this project)
> 2. Fixed ✅ (i was picking the wrong ones from the theme)
3. It's from the existing theme primary color, which on these do you prefer from the images in `images preview`? 
<details><summary>images preview</summary>
<p>

|                               Grey                          |                          Light Blue                          |
|--------------------------------------|-----------------------------------------|
| <img width="628" alt="Screenshot 2023-05-20 at 02 45 30" src="https://github.com/mkobuolys/flutter-design-patterns/assets/59662912/50704909-b0a8-4e79-9934-d501b5023d09"> | <img width="634" alt="Screenshot 2023-05-20 at 02 43 34" src="https://github.com/mkobuolys/flutter-design-patterns/assets/59662912/87859e5e-5fe5-4205-a455-3856207f78cb"> | 

</p>
</details> 

> 4.1 Fixed ✅ 
> 4.2 Fixed ✅  (In this case there's no need, there's was a use-case in a project of mine but it's not present here)
> 4.3 Fixed ✅
> 4.5 Fixed ✅ 

4.6 Let me know if the change here is fine, (moved to a `helpers/dart_syntax_highlighter.dart`)

_extra: we can get rid of `parseSyntaxLanguage` since it's dart only syntax now_

### Current state
<img width="1194" alt="Screenshot 2023-05-20 at 03 40 16" src="https://github.com/mkobuolys/flutter-design-patterns/assets/59662912/5d728df4-1406-48a0-abe0-91db13c09bfb">

mkobuolys commented:
@Juliotati Sorry for the noise (not sure if you got a notification) - PR is currently waiting for the code changes.

Juliotati commented:
@mkobuolys  I'm a bit lost on what you mean by, "waiting for code changes".
Is there something missing that i missed?

Also please check my reply above on point `3.`.
Just need you to let me now which it better, Grey or Light Blue?

mkobuolys commented:
> @mkobuolys  I'm a bit lost on what you mean by, "waiting for code changes".
> 
> Is there something missing that i missed?
> 
> 
> 
> Also please check my reply above on point `3.`.
> 
> Just need you to let me now which it better, Grey or Light Blue?

Check the PR - I've left a bunch of comments there. For the 3rd point, grey looks better to me.

Juliotati commented:
> Check the PR (files) - I've left a bunch of comments there. For the 3rd point, grey looks better to me.

Oh the comments won't be visible on my side until you do "request changes"

mkobuolys commented:
> > Check the PR (files) - I've left a bunch of comments there. For the 3rd point, grey looks better to me.
> 
> Oh the comments won't be visible on my side until you do "request changes". I've had that issue before on GitHub too.

Ok, I thought you see them - my bad.

Juliotati commented:
haha no issue, i thought you were busy so i was patiently waiting for the feedback.
Thanks i'll take a look and clean up ✨

Juliotati commented on lib/helpers/dart_sytax_highlighter.dart:
```diff
@@ -0,0 +1,94 @@
+
+  List<TextSpan> _convert(List<Node> nodes) {
+    final List<TextSpan> spans = [];
+    var currentSpans = spans;
```
Yes, but in this case item added to `currentSpans` are present in `spans` too.
Creating and assigning a new list won't work (_result is empty_) because the lists need to be connected.
But it's not needed actually. I'll get rid of `spans` and only keep `currentSpans`, as there's the same result as before.

Juliotati commented on lib/modules/design_pattern_details/widgets/rich_markdown_body.dart:
```diff
@@ -0,0 +1,109 @@
+          color: textTheme.bodyMedium?.color?.withOpacity(0.8),
+        ),
+        tableHead: bodyMedium?.copyWith(
+          fontSize: fontSize + 2,
```
It's mainly for visual difference. To have table headers always be 2 points up from the defined `fontSize`.
I can keep it the same instead of calculating a new size there if preferred. 
What's your view on it? @mkobuolys

mkobuolys replied:
Fair, we can leave it this way.

Juliotati commented on lib/modules/design_pattern_details/widgets/rich_markdown_body.dart:
```diff
@@ -0,0 +1,109 @@
+        blockquote: bodyMedium?.copyWith(
+          color: theme.colorScheme.onSurface.withOpacity(0.6),
+        ),
+        code: bodyMedium?.copyWith(
```
@mkobuolys  From the configurations available, it's not possible mainly because `code` doesn't have decoration yet, only has text style available for customization.
I tend to use a different font for code and codeBlock, namely `FiraCode` for a better visual difference.
But i'm not sure you want that at the moment.

The closest i've gotten to the image you shared is on the screenshot below:

- `code` and `codeBlock` font is `FiraCode`
- `code` background is highlighted _(tried to apply the same on your project but, not happening, don't know why yet)_
- 🚨 An issue is that when you select the text you can't see that `code` was selected, because the selection is behind the code highlight

_screenshot from a project of mine:_
<img width="501" alt="Screenshot 2023-05-28 at 01 17 09" src="https://github.com/mkobuolys/flutter-design-patterns/assets/59662912/90c5b612-affc-4ceb-b12b-231b3e16ba0c">

Juliotati replied:
Posted info here -> https://github.com/mkobuolys/flutter-design-patterns/pull/52#issuecomment-1565771800

mkobuolys replied:
Replied to your comment, check it in the PR.

Juliotati commented:
> Is it possible to highlight the code like this (or at least similar to this)?
> 
> <img alt="Screenshot 2023-05-21 at 02 36 12" width="739" src="https://user-images.githubusercontent.com/17415795/239713294-0f0b7e13-2c88-41de-bae1-d3011eedab4b.png">

For your re quest here, looks like the only way is to edit `code` text `background` from its `textStyle`.
Here's the current status, not sure if i'm missing something at the stroke is not being rounded. 

| Preview | Code |
|--------|--------|
| ![Screenshot 2023-05-28 at 01 56 30](https://github.com/mkobuolys/flutter-design-patterns/assets/59662912/51556ad8-2409-4a15-81ab-a102b08ce97f) | ![Screenshot 2023-05-28 at 01 55 04](https://github.com/mkobuolys/flutter-design-patterns/assets/59662912/3f93188f-42d9-407a-8bdd-f683eeb89629)|

Juliotati commented:
### Code Font

Any preference?
| Normal weight | w700 |
|--------|--------|
| <img width="674" alt="Screenshot 2023-05-28 at 02 44 33" src="https://github.com/mkobuolys/flutter-design-patterns/assets/59662912/06a7ec06-07b0-4b89-a1c0-3a369eb6d053"> | <img width="676" alt="Screenshot 2023-05-28 at 02 43 11" src="https://github.com/mkobuolys/flutter-design-patterns/assets/59662912/946e78bd-b210-4755-9aaf-d64378501983"> |

Juliotati reviewed (commented):
ready for another review

mkobuolys commented:
> > Is it possible to highlight the code like this (or at least similar to this)?
> > <img alt="Screenshot 2023-05-21 at 02 36 12" width="739" src="https://user-images.githubusercontent.com/17415795/239713294-0f0b7e13-2c88-41de-bae1-d3011eedab4b.png">
> 
> ## Code background
> For your request, looks like the only way is to edit `code` `background` from its `textStyle`.
> 
> Preview	Code
> ![Screenshot 2023-05-28 at 01 56 30](https://user-images.githubusercontent.com/59662912/241458150-51556ad8-2409-4a15-81ab-a102b08ce97f.png)	![Screenshot 2023-05-28 at 01 55 04](https://user-images.githubusercontent.com/59662912/241458104-3f93188f-42d9-407a-8bdd-f683eeb89629.png)
> 🚨 Apparently the rounded strokes defined here are only visible when stroke width has some high value, eg. 10+, which just looks wrong ☠️ 🚨 Manually controlling the radius here is not possible somehow.
> 
> ![Screenshot 2023-05-28 at 02 11 09](https://user-images.githubusercontent.com/59662912/241460395-f29c0257-547e-475a-bc5e-c1ebac11dc72.png)
> 
> ## Alternative
> <img alt="Screenshot 2023-05-28 at 02 54 47" width="670" src="https://user-images.githubusercontent.com/59662912/241465174-7323eadd-4e41-440b-a02b-8b9db029be20.png">

I played with the code style parameters, found a good enough option:

```
code: bodyMedium?.copyWith(
  color: theme.colorScheme.onSurface,
  backgroundColor: theme.colorScheme.surface,
),
```

mkobuolys commented:
> ### Code Font `FiraCode`
> Any preference?
> 
> Normal weight	w700
> <img alt="Screenshot 2023-05-28 at 02 44 33" width="674" src="https://user-images.githubusercontent.com/59662912/241464347-06a7ec06-07b0-4b89-a1c0-3a369eb6d053.png">	<img alt="Screenshot 2023-05-28 at 02 43 11" width="676" src="https://user-images.githubusercontent.com/59662912/241464237-946e78bd-b210-4755-9aaf-d64378501983.png">

No need for a font change - it looks fine now. Don't want to blindly change to a random one.

mkobuolys commented on lib/helpers/dart_syntax_highlighter.dart:
```diff
@@ -0,0 +1,97 @@
+import 'package:flutter/cupertino.dart';
```
??? Just use `BuildContext` from the material package.

mkobuolys commented on lib/modules/design_pattern_details/widgets/rich_markdown_body.dart:
```diff
@@ -0,0 +1,98 @@
+        p: bodyMedium,
+        blockSpacing: LayoutConstants.paddingM,
+        blockquotePadding: const EdgeInsets.symmetric(
+          horizontal: 12,
```
```suggestion
          horizontal: LayoutConstants.paddingM * 1.5,
```

mkobuolys commented on lib/modules/design_pattern_details/widgets/rich_markdown_body.dart:
```diff
@@ -0,0 +1,98 @@
+        h5: textTheme.titleMedium,
+        h6: textTheme.titleLarge,
+        p: bodyMedium,
+        blockSpacing: LayoutConstants.paddingM,
```
```suggestion
        blockSpacing: LayoutConstants.spaceM,
```

mkobuolys commented on lib/modules/design_pattern_details/widgets/rich_markdown_body.dart:
```diff
@@ -0,0 +1,98 @@
+      ),
+      onTapLink: (_, link, __) => UrlLauncher.launchUrl(link ?? ''),
+      syntaxHighlighter: DartSyntaxHighlighter(context),
+      data: data ??'',
```
```suggestion
      data: data ?? '',
```

mkobuolys commented on lib/helpers/dart_syntax_highlighter.dart:
```diff
@@ -0,0 +1,97 @@
+
+    void traverse(Node node) {
+      if (node.value != null) {
+        currentSpans.add(
+          node.className == null
+              ? TextSpan(text: node.value)
+              : TextSpan(text: node.value, style: _syntaxTheme[node.className]),
+        );
```
```suggestion
        currentSpans.add(
          TextSpan(
            text: node.value,
            style: node.className != null ? _syntaxTheme[node.className] : null
          ),
        );
```

mkobuolys replied:
Still, not really resolved since only the style property is different, I would put the conditional logic there. Anyway, not a biggie.

Juliotati replied:
oh, my bad i didn't notice the `null` check you added. I thought it was only line formatting

mkobuolys commented on lib/modules/design_pattern_details/widgets/rich_markdown_body.dart:
```diff
@@ -0,0 +1,98 @@
+
+    final textTheme = theme.textTheme;
+    final bodyMedium = textTheme.bodyMedium;
+    final fontSize = theme.textTheme.bodyMedium?.fontSize ?? 14.0;
```
```suggestion
    final fontSize = bodyMedium?.fontSize ?? 14.0;
```

Juliotati commented:
> I played with the code style parameters, and found a good enough option:
>
> code: bodyMedium?.copyWith(
>   color: theme.colorScheme.onSurface,
>   backgroundColor: theme.colorScheme.surface,
> ),
> Just the code highlighting does not work - not sure what's missing.

The highlight is there but it's behind the backgroundColor.
This should make it a little bit visible but it's at the cost of the `code` being less pronounced

```dart
code: bodyMedium?.copyWith(
   color: theme.colorScheme.onSurface,
   backgroundColor: theme.colorScheme.surface.withOpacity(0.6),
),
```

mkobuolys reviewed (approved):
LGTM, thanks for your contribution! ✌️
